src/rdkafka_broker.c

```diff
@@ -1567,6 +1569,111 @@ int rd_kafka_socket_cb_generic (int domain, int type, int protocol,
 }
 
 
+
+/**
+ * @brief Update the reconnect backoff.
+ *        Should be called when a connection is made.
+ *
+ * @locality broker thread
+ * @locks none
+ */
+static void
+rd_kafka_broker_update_reconnect_backoff (rd_kafka_broker_t *rkb,
+                                          const rd_kafka_conf_t *conf,
+                                          rd_ts_t now) {
+        int backoff;
+
+        /* If last connection attempt was more than reconnect.backoff.max.ms
+         * ago, reset the reconnect backoff to the initial
+         * reconnect.backoff.ms value.
+         * Or if the current backoff increment has reached the max value. */
+        if (rkb->rkb_ts_reconnect + (conf->reconnect_backoff_max_ms * 1000) <
+            now ||
+            rkb->rkb_reconnect_backoff_ms >= conf->reconnect_backoff_max_ms)
+                rkb->rkb_reconnect_backoff_ms = conf->reconnect_backoff_ms;
+
+        /* Apply +-20% jitter to next backoff. */
+        backoff = rd_jitter((int)(float)rkb->rkb_reconnect_backoff_ms * 0.8,
+                            (int)(float)rkb->rkb_reconnect_backoff_ms * 1.2);
+
+         /* Cap to reconnect.backoff.max.ms. */
+        backoff = RD_MIN(backoff, conf->reconnect_backoff_max_ms);
+
+        /* Set time of next reconnect */
+        rkb->rkb_ts_reconnect = now + (backoff * 1000);
```

Comment:
Can we name this rkb_ts_reconnect_us to make it more clear that this is in microseconds?  Otherwise, this is kind of confusing.

Reply:
Good point, but the `ts_` prefix convention is used throughout the code to indicate a microsecond rd_ts_t absolute timestamp:

```
$ git grep 'rd_ts_t.*ts_.*;' src/*.h
src/rdinterval.h:        rd_ts_t    ri_ts_last; /* last interval timestamp */
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_rsal_last;
src/rdkafka_broker.h:   rd_ts_t             rkb_ts_fetch_backoff;
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_state;        /* Timestamp of last
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_tx_last;    /**< Timestamp of last
src/rdkafka_broker.h:   rd_ts_t             rkb_ts_metadata_poll; /* Next metadata poll time */
src/rdkafka_broker.h:        rd_ts_t             rkb_ts_reconnect;
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_retry;    /* Absolute send retry time */
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_enq;
src/rdkafka_buf.h:      rd_ts_t rkbuf_ts_sent;    /* Initially: Absolute time of transmission,
src/rdkafka_buf.h:        rd_ts_t rkbuf_ts_timeout; /* Request timeout (absolute time). */
src/rdkafka_cgrp.h:        rd_ts_t            rkcg_ts_statechange;     /* Timestamp of last
src/rdkafka_cgrp.h:     rd_ts_t            rkcg_ts_terminate;       /* Timestamp of when
src/rdkafka_cgrp.h:                rd_ts_t            ts_rebalance;       /* Timestamp of
src/rdkafka_int.h:        rd_ts_t          rk_ts_metadata;    /* Timestamp of most recent
src/rdkafka_int.h:      rd_ts_t          rk_ts_full_metadata;       /* Timesstamp of .. */
src/rdkafka_int.h:                rd_ts_t ts_idemp_state;/**< Last state change */
src/rdkafka_metadata.h:        rd_ts_t rkmce_ts_expires;                /* Expire time */
src/rdkafka_metadata.h:        rd_ts_t rkmce_ts_insert;                 /* Insert time */
src/rdkafka_msg.h:                        rd_ts_t ts_timeout; /* Message timeout */
src/rdkafka_msg.h:                        rd_ts_t ts_enq;     /* Enqueue/Produce time */
src/rdkafka_msg.h:                        rd_ts_t ts_backoff; /* Backoff next Produce until
src/rdkafka_op.h:                        rd_ts_t ts_timeout;
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_fetch_backoff; /* Back off fetcher for
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_committed_offset; /* Timestamp of last
src/rdkafka_partition.h:        rd_ts_t            rktp_ts_offset_lag;
src/rdkafka_topic.h:    rd_ts_t            rkt_ts_metadata; /* Timestamp of last metadata
```